test/storage/hash_index_test.cpp

```diff
@@ -0,0 +1,1645 @@
+    // previously committed versions
+    if (worker_id % 2 == 0) {
+      for (uint32_t i = 0; i < num_inserts_; i++) {
+        auto *const insert_txn = txn_manager_.BeginTransaction();
+        auto *const insert_redo =
+            insert_txn->StageWrite(CatalogTestUtil::test_db_oid, CatalogTestUtil::test_table_oid, tuple_initializer_);
+        auto *const insert_tuple = insert_redo->Delta();
+        *reinterpret_cast<int32_t *>(insert_tuple->AccessForceNotNull(0)) = i;
+        const auto tuple_slot = sql_table_->Insert(insert_txn, insert_redo);
+
+        *reinterpret_cast<int32_t *>(insert_key->AccessForceNotNull(0)) = i;
+        if (unique_index_->InsertUnique(insert_txn, *insert_key, tuple_slot)) {
+          txn_manager_.Commit(insert_txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+        } else {
+          txn_manager_.Abort(insert_txn);
+        }
+      }
+
+    } else {
+      for (uint32_t i = num_inserts_ - 1; i < num_inserts_; i--) {
+        auto *const insert_txn = txn_manager_.BeginTransaction();
+        auto *const insert_redo =
+            insert_txn->StageWrite(CatalogTestUtil::test_db_oid, CatalogTestUtil::test_table_oid, tuple_initializer_);
+        auto *const insert_tuple = insert_redo->Delta();
+        *reinterpret_cast<int32_t *>(insert_tuple->AccessForceNotNull(0)) = i;
+        const auto tuple_slot = sql_table_->Insert(insert_txn, insert_redo);
+
+        *reinterpret_cast<int32_t *>(insert_key->AccessForceNotNull(0)) = i;
+        if (unique_index_->InsertUnique(insert_txn, *insert_key, tuple_slot)) {
+          txn_manager_.Commit(insert_txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+        } else {
+          txn_manager_.Abort(insert_txn);
+        }
+      }
+    }
+    delete[] key_buffer;
+  };
+
+  // run the workload
+  for (uint32_t i = 0; i < num_threads_; i++) {
+    thread_pool_.SubmitTask([i, &workload] { workload(i); });
+  }
+  thread_pool_.WaitUntilAllFinished();
+
+  // scan the results
+  auto *const scan_txn = txn_manager_.BeginTransaction();
+
+  std::vector<storage::TupleSlot> results;
+
+  auto *const key_pr = default_index_->GetProjectedRowInitializer().InitializeRow(key_buffer_1_);
+
+  // scan[0,num_inserts_) should hit num_inserts_ keys (no duplicates)
+  for (uint32_t i = 0; i < num_inserts_; i++) {
+    *reinterpret_cast<int32_t *>(key_pr->AccessForceNotNull(0)) = i;
+    unique_index_->ScanKey(*scan_txn, *key_pr, &results);
+    EXPECT_EQ(results.size(), 1);
+    results.clear();
+  }
+
+  txn_manager_.Commit(scan_txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+}
+
+/**
+ * This test creates multiple worker threads that all try to insert [0,num_inserts) as tuples in the table and into the
+ * primary key index. At completion of the workload, all num_inserts_ txns * num_threads_ should have committed with
+ * visible versions in the index and table.
+ */
+// NOLINTNEXTLINE
+TEST_F(HashIndexTests, DefaultInsert) {
```

Comment:
Side effect of our index ConstraintType currently being DEFAULT and UNIQUE. We might need to get the interested parties in a room (not many people right now) to codify where our index metadata is living and the terminology being used, since all of this pre-dates the Postgres catalog.